framework/src/constraints/NodeElemConstraint.C

```diff
@@ -0,0 +1,274 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NodeElemConstraint.h"
+
+// MOOSE includes
+#include "Assembly.h"
+#include "MooseEnum.h"
+#include "MooseMesh.h"
+#include "MooseVariableFE.h"
+#include "SystemBase.h"
+
+#include "libmesh/string_to_enum.h"
+
+template <>
+InputParameters
+validParams<NodeElemConstraint>()
+{
+  MooseEnum orders("FIRST SECOND THIRD FOURTH", "FIRST");
+  InputParameters params = validParams<Constraint>();
+  params.addRequiredParam<SubdomainName>("slave", "slave block id");
+  params.addRequiredParam<SubdomainName>("master", "master block id");
+  params.addParam<MooseEnum>("order", orders, "The finite element order used for projections");
+  params.addRequiredCoupledVar("master_variable", "The variable on the master side of the domain");
+
+  return params;
+}
+
+NodeElemConstraint::NodeElemConstraint(const InputParameters & parameters)
+  : Constraint(parameters),
+    // The slave side is at nodes (hence passing 'true').  The neighbor side is the master side and
+    // it is not at nodes (so passing false)
+    NeighborCoupleableMooseVariableDependencyIntermediateInterface(this, true, false),
+    NeighborMooseVariableInterface<Real>(
+        this, true, Moose::VarKindType::VAR_NONLINEAR, Moose::VarFieldType::VAR_FIELD_STANDARD),
+
+    _slave(_mesh.getSubdomainID(getParam<SubdomainName>("slave"))),
+    _master(_mesh.getSubdomainID(getParam<SubdomainName>("master"))),
+
+    _master_q_point(_assembly.qPoints()),
+    _master_qrule(_assembly.qRule()),
+
+    _current_node(_var.node()),
+    _current_elem(_var.neighbor()),
+
+    _u_slave(_var.dofValues()),
+    _u_slave_old(_var.dofValuesOld()),
+    _phi_slave(1),
+    _test_slave(1), // One entry
+
+    _master_var(*getVar("master_variable", 0)),
+    _master_var_num(_master_var.number()),
+
+    _phi_master(_assembly.phiNeighbor(_master_var)),
+    _grad_phi_master(_assembly.gradPhiNeighbor(_master_var)),
+
+    _test_master(_var.phiNeighbor()),
+    _grad_test_master(_var.gradPhiNeighbor()),
+
+    _u_master(_master_var.slnNeighbor()),
+    _u_master_old(_master_var.slnOldNeighbor()),
+    _grad_u_master(_master_var.gradSlnNeighbor()),
+
+    _dof_map(_sys.dofMap()),
+    _node_to_elem_map(_mesh.nodeToElemMap()),
+
+    _overwrite_slave_residual(true)
+{
+  _mesh.errorIfDistributedMesh("NodeElemConstraint");
+
+  addMooseVariableDependency(&_var);
+  // Put a "1" into test_slave
+  // will always only have one entry that is 1
+  _test_slave[0].push_back(1);
+
+  // get mesh pointLocator
+  std::unique_ptr<PointLocatorBase> pointLocator = _mesh.getPointLocator();
+  pointLocator->enable_out_of_mesh_mode();
+  const std::set<subdomain_id_type> allowed_subdomains{_master};
+
+  // slave id and master id
+  dof_id_type sid, mid;
+
+  // prepare _slave_to_master_map
```

Comment:
It seems to me like this code should belong in the derived class because in general, the NodeElemConstraint could be used to enforce constraints that are not based on locating nodes within elements.

Reply:
I'll write a prepareSlaveToMasterMap() method and make it abstract in the base class. Then EqualValueEmbeddedConstraint will override this method.